<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <arg name="ns" default=""/>
  <arg name="rviz" default="true"/>
  <arg name="base_frame"/>
  <arg name="lidar_frame"/>
  <arg name="pointcloud_topic"/>
  <arg name="calibration_service_name" default="extrinsic_calibration"/>

  <let name="rviz_profile" value="$(find-pkg-share ground_plane_calibrator)/rviz/default.rviz"/>

  <arg name="max_inlier_distance" default="0.03"/>
  <arg name="min_plane_points" default="500"/>
  <arg name="min_plane_points_percentage" default="20.0"/>
  <arg name="max_cos_distance" default="0.2"/>
  <arg name="max_iterations" default="500"/>
  <arg name="verbose" default="true"/>
  <arg name="overwrite_xy_yaw" default="false"/>
  <arg name="filter_estimations" default="true"/>

  <arg name="use_crop_box_filter" default="true"/>
  <arg name="crop_box_min_x" default="-50.0"/>
  <arg name="crop_box_min_y" default="-50.0"/>
  <arg name="crop_box_min_z" default="-50.0"/>
  <arg name="crop_box_max_x" default="50.0"/>
  <arg name="crop_box_max_y" default="50.0"/>
  <arg name="crop_box_max_z" default="50.0"/>
  <arg name="use_pca_rough_normal" default="false"/>

  <node pkg="tf2_ros" exec="static_transform_publisher" name="initial_base_link_broadcaster" output="screen" args="0 0 0 0 0 0 $(var base_frame) estimated_base_link"/>
  <node pkg="tf2_ros" exec="static_transform_publisher" name="lidar_broadcaster" output="screen" args="0 0 0 0 0 0 $(var lidar_frame) lidar_frame"/>

  <group>
    <push-ros-namespace namespace="$(var ns)"/>
    <node pkg="ground_plane_calibrator" exec="ground_plane_calibrator" name="ground_plane_calibrator" output="screen">
      <remap from="input_pointcloud" to="$(var pointcloud_topic)"/>
      <remap from="extrinsic_calibration" to="$(var calibration_service_name)"/>

      <param name="base_frame" value="$(var base_frame)"/>
      <param name="lidar_frame" value="$(var lidar_frame)"/>
      <param name="marker_size" value="10.0"/>
      <param name="max_inlier_distance" value="$(var max_inlier_distance)"/>
      <param name="min_plane_points" value="$(var min_plane_points)"/>
      <param name="min_plane_points_percentage" value="$(var min_plane_points_percentage)"/>
      <param name="max_cos_distance" value="$(var max_cos_distance)"/>
      <param name="max_iterations" value="$(var max_iterations)"/>
      <param name="verbose" value="$(var verbose)"/>
      <param name="overwrite_xy_yaw" value="$(var overwrite_xy_yaw)"/>
      <param name="filter_estimations" value="$(var filter_estimations)"/>

      <param name="use_crop_box_filter" value="$(var use_crop_box_filter)"/>
      <param name="crop_box_min_x" value="$(var crop_box_min_x)"/>
      <param name="crop_box_min_y" value="$(var crop_box_min_y)"/>
      <param name="crop_box_min_z" value="$(var crop_box_min_z)"/>
      <param name="crop_box_max_x" value="$(var crop_box_max_x)"/>
      <param name="crop_box_max_y" value="$(var crop_box_max_y)"/>
      <param name="crop_box_max_z" value="$(var crop_box_max_z)"/>
      <param name="use_pca_rough_normal" value="$(var use_pca_rough_normal)"/>

      <param name="initial_angle_cov" value="5.0"/>
      <!--degrees -->
      <param name="initial_translation_cov" value="0.05"/>
      <!--meters -->
      <param name="angle_measurement_cov" value="0.5"/>
      <!--degrees -->
      <param name="angle_process_cov" value="0.1"/>
      <!--degrees -->
      <param name="translation_measurement_cov" value="0.005"/>
      <!--meters -->
      <param name="translation_process_cov" value="0.001"/>
      <!--meters -->
      <param name="angle_convergence_threshold" value="0.02"/>
      <!--degrees -->
      <param name="translation_convergence_threshold" value="0.0005"/>
      <!--meters -->
    </node>
  </group>

  <node pkg="rviz2" exec="rviz2" name="rviz2" output="screen" args="-d $(var rviz_profile)">
    <remap from="pointcloud" to="$(var pointcloud_topic)"/>
  </node>
</launch>
